### DISCUSSION TITLE
Feature: Custom Functions and Formats

### DISCUSSION THREAD

**Discussion by kazad** [kazad/instacalc#110] [Announcements]
You can now define custom functions (and formats). Define functions similar to JS, or use math-style `f(x) = ... `

```
function f(x, y) { 2x + 3y }
f(x, y) = 2x + 3y
```
The last result is returned automatically. Early return isn't yet possible (`return = 5%` is a common variable name, and I don't want the keyword to take precendence).

Call functions as you'd expect:  `f(2, 3)`

Demo: https://instacalc.com/57811

```
Instacalc: Custom functions and formats

custom function
function circlearea(r) { pi * r^2 }                         = [object Object]
circlearea(10)                                              = 314.15926536
circlearea(10 cm)                                           = 314.1593 cm²
10 cm -> circlearea -> in                                   = 48.6948 in²

custom format (really, a function behind the scenes)
function fmt(x) { if (x > 1000) { x as eng} else {x} }      = [object Object]
fmt(500)                                                    = 500
fmt(5000)                                                   = 5e3
500 as fmt                                                  = 500
5000 as fmt                                                 = 5e3
```

A function can be invoked as a conversion or format endpoint, such as `15 cm as f` or `15 cm -> f`.

I've always liked the "pipeline" style of function calling `10 cm -> circlearea -> in`, which matches how I think. An input is sent through various steps, and the result pops out.

**Comment by SamHasler** [comment]
Can functions take a range? How would I write one to e.g.: sort a range numerically and return the penultimate element (i.e. the second largest value in the range)

**Reply by kazad** [reply]
Great question! Yep, you can pass in a range like `sort(r2:r4)`, I just added sorting functions.

Demo: https://instacalc.com/58757

In one line, you can do `(sort(r2:r4) -> reverse)[1]` to get the 2nd largest item (arrays are zero-indexed).
